AlphaWallet/Core/CoinTicker/CoinTickersFetcher.swift

```diff
@@ -0,0 +1,310 @@
+            self.server = server
+        }
+    }
+
+    func fetchCoinTickers(for enabledObjects: [TokenMappedToTicker]) {
+        fetchCoinTickersPromise(for: enabledObjects).done { [weak self] tickers in
+            self?.tickers = tickers
+
+            if !tickers.isEmpty {
+                self?.fetchChartHistoryForAlearyResolvedTickers()
+            }
+        }.catch { _ in
+            //no-op
+        }
+    }
+
+    private static let fetchChartHistoryTimeLimit: TimeInterval = 5
+    private let loadChartHistoryLimiter = TimedLimiter(limit: CoinTickersFetcher.fetchChartHistoryTimeLimit)
+
+    func fetchChartHistoryForAlearyResolvedTickers() {
+        loadChartHistoryLimiter.execute { [weak self] in
+            guard let strongSelf = self else { return }
+            guard let value = strongSelf.tickers.first(where: { strongSelf.history[$0.value] == nil }) else { return }
+
+            strongSelf.loadAllPeriodsHistory(for: value.key).done { _ in
+                CoinTickersFetcher.queue.asyncAfter(deadline: .now() + Self.fetchChartHistoryTimeLimit + 1) {
+                    strongSelf.fetchChartHistoryForAlearyResolvedTickers()
+                }
+            }.cauterize()
+        }
+    }
+
+    private func loadAllPeriodsHistory(for key: AddressToPRCServerKey) -> Promise<Void> {
+        return Promise<Void> { seal in
+            let group = DispatchGroup()
+            for period in ChartHistoryPeriod.allCases {
+                group.enter()
+
+                fetchChartHistory(force: true, period: period, for: key).done { _ in
+                    //no-op
+                }.catch { _ in
+                    //no-op
+                }.finally {
+                    group.leave()
+                }
+            }
+
+            group.notify(queue: CoinTickersFetcher.queue) {
+                seal.fulfill(())
+            }
+        }
+    }
+
+    private func loadChartHistory(period: ChartHistoryPeriod, ticker: CoinTicker) -> Promise<ChartHistory> {
+        firstly {
+            provider.request(.charts(config: config, id: ticker.id, currency: Constants.Currency.usd, days: period.rawValue))
+        }.map(on: CoinTickersFetcher.queue, { response -> ChartHistory in
+            return try response.map(ChartHistory.self, using: JSONDecoder())
+        }).recover(on: CoinTickersFetcher.queue, { _ -> Promise<ChartHistory> in
+            return .value(.empty)
+        })
+    }
+
+    private func cacheChartHistory(result: ChartHistory, period: ChartHistoryPeriod, for ticker: CoinTicker) -> Promise<ChartHistory> {
+        guard !result.prices.isEmpty else { return .value(result) }
+
+        var valueToSet = self.history[ticker] ?? [:]
+        valueToSet[period] = result
+
+        self.history[ticker] = valueToSet
+        return .value(result)
+    }
+
+    func fetchChartHistory(force: Bool, period: ChartHistoryPeriod, for key: AddressToPRCServerKey) -> Promise<ChartHistory> {
+        return firstly {
```

Comment:
Can you describe how the fetching and caching works? I'm especially interested if we are caching and especially for data that is not `ChartHistoryPeriod.day` and when/how we purge the cache.

Caching aside, do we make 5 calls per token? (1 for each case of `ChartHistoryPeriod`)?

I reckon the API rate limits by IP address, so in an event, like a conference, we might only have a very small set of IP addresses with several tends or hundreds of devices :)

Reply:
`Caching aside, do we make 5 calls per token? (1 for each case of ChartHistoryPeriod)?` yes, we make api calls to fetch history values.  one time per 5 seconds for single server. Once it done we cache fetched values in dictionary.

Reply:
but, seems like its need to be updated. because when we have 5 or more enabled networks it will cause to many requests.